x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java

```diff
@@ -588,6 +620,76 @@ private LogicalPlan resolveLookup(Lookup l, List<Attribute> childrenOutput) {
+                    String name = cols.get(0).name();
+                    UnresolvedAttribute errorAttribute = new UnresolvedAttribute(
+                        join.source(),
+                        name,
+                        "Only LEFT join is supported with USING"
+                    );
+                    return join.withConfig(new JoinConfig(type, singletonList(errorAttribute), emptyList(), emptyList()));
+                }
+                // resolve the using columns against the left and the right side then assemble the new join config
+                List<Attribute> leftKeys = resolveUsingColumns(cols, join.left().output(), "left");
+                List<Attribute> rightKeys = resolveUsingColumns(cols, join.right().output(), "right");
+                List<Attribute> output = new ArrayList<>(join.left().output());
+                output.addAll(join.right().outputSet().subtract(new AttributeSet(rightKeys)));
+
+                // update the config - pick the left keys as those in the output
+                type = new UsingJoinType(coreJoin, rightKeys);
+                config = new JoinConfig(type, leftKeys, leftKeys, rightKeys);
+                join = new LookupJoin(join.source(), join.left(), join.right(), config, output);
+            }
+            // everything else is unsupported for now
+            else {
+                UnresolvedAttribute errorAttribute = new UnresolvedAttribute(join.source(), "unsupported", "Unsupported join type");
+                // add error message
+                return join.withConfig(new JoinConfig(type, singletonList(errorAttribute), emptyList(), emptyList()));
+            }
+            return join;
+        }
+
+        private List<Attribute> resolveUsingColumns(List<Attribute> cols, List<Attribute> output, String side) {
+            List<Attribute> resolved = new ArrayList<>(cols.size());
+            for (Attribute col : cols) {
+                if (col instanceof UnresolvedAttribute ua) {
+                    Attribute resolvedCol = maybeResolveAttribute(ua, output);
+                    if (resolvedCol instanceof UnresolvedAttribute ucol) {
+                        String message = ua.unresolvedMessage();
+                        String match = "column [" + ucol.name() + "]";
+                        resolvedCol = ucol.withUnresolvedMessage(message.replace(match, match + "in " + side + " side of join"));
+                    }
+                    resolved.add(resolvedCol);
+                }
+                // columns are expected to be unresolved - if that's not the case return an error
+                else {
+                    return singletonList(
+                        new UnresolvedAttribute(
+                            col.source(),
+                            col.name(),
+                            "Surprised to discover column [ " + col.name() + "] already resolved"
+                        )
+                    );
+                }
```

Comment:
I think this should be an IllegalStateException - otherwise we mask an internal server error (500) as an invalid query (400).

Reply:
Added item to #116208